### ISSUE TITLE
Adagrad colocation fitting error

### ISSUE THREAD

**Issue by kkranen** [tensorflow/tensorflow#41777]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 16.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: N/A
- TensorFlow installed from (source or binary): Binary
- TensorFlow version (use command below): 2.2.0
- Python version: 3.6.5
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: 11.0
- GPU model and memory: Nvidia Quadro V100

**Describe the current behavior**
Compilation and fitting of a network that uses tf.keras.layers.DenseFeatures and an Adagrad optimizer results in a co-location error that does not resolve when allowing soft placement and/or turning off XLA JIT. Changing the optimizer to SGD resolves the issue.

**Describe the expected behavior**
Adagrad should not incite a colocation error at model fitting when used with DenseFeatures.

**Standalone code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.

```python
import tensorflow as tf
from tensorflow.python.feature_column import feature_column_v2 as fc
metadata = {'m1': tf.ones(shape=(100,1)), 'm2': tf.ones(shape=(100,1)),'label':tf.ones(shape=(100,1))}
num_samples = 100
dnn_optimizer = tf.keras.optimizers.Adagrad(
                learning_rate=1
            )
def meta_dict_gen():
    for i in range(num_samples):
        ls = {}
        for key, val in metadata.items():
            ls[key] = val[i]
        yield ls
# DATASET CREATION
d = tf.data.Dataset.from_generator(
    meta_dict_gen,
    output_types={k: tf.float32 for k in metadata},
    output_shapes={'m1': (1,), 'm2': (1),'label':(1)})
d = d.shuffle(
        buffer_size=10 * 8
    )


features = {'m1':1,'m2':1}
def label_map(d):
    
    label = d.pop('label')
    reshaped_label = tf.reshape(label, [-1, label.shape[-1]])
    reshaped_elem = {
        key: tf.reshape(d[key], [-1, d[key].shape[-1]])
        for key in d if key in features.keys()
    }
    
    return reshaped_elem, reshaped_label
d = d.map(map_func=label_map)

# CREATING DENSE FEATURE LAYER
d_columns = [tf.feature_column.embedding_column(fc.categorical_column_with_hash_bucket(key='m1', hash_bucket_size=2, dtype=tf.int64),dimension=1,combiner='mean'),
tf.feature_column.numeric_column(
                    'm2', shape=(1,))]

d_features = {}
d_features['m1'] = tf.keras.Input(shape=(1,), name='m1', dtype=tf.int64, sparse=False)
d_features['m2'] = tf.keras.Input(shape=(1,), name='m2', dtype=tf.int64, sparse=False)


#CREATING MODEL

d_input = tf.keras.layers.DenseFeatures(d_columns, name='d_embedded')(d_features)
d_output = tf.keras.layers.Dense(1)(d_input)
d_model = tf.keras.Model(d_features,d_output)
d_model.compile()
d_model.compile(optimizer = dnn_optimizer,loss= 'binary_crossentropy',metrics = ['binary_crossentropy'])
d_model.fit(d)
```

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

@DEKHTIARJonathan @nluehr

**Comment by Saduf2019** [comment]
@kkranen 
I ran the code shared and did not face any error, please find the [gist here](https://colab.research.google.com/gist/Saduf2019/65890558a70bd7bb814c9b7250af061f/untitled300.ipynb).
PLease sharea  colab gist witht the error and the error logs for us to analyse.

**Comment by kkranen** [comment]
Hmm, I see the same results you achieved with colab on my side as well. Locally, my stack trace looks like this when I attempt to run the block:


```
---------------------------------------------------------------------------
InvalidArgumentError                      Traceback (most recent call last)
<ipython-input-1-476fdc9b6732> in <module>
     52 d_model.compile()
     53 d_model.compile(optimizer = dnn_optimizer,loss= 'binary_crossentropy',metrics = ['binary_crossentropy'])
---> 54 d_model.fit(d)

/usr/local/lib/python3.6/dist-packages/tensorflow/python/keras/engine/training.py in _method_wrapper(self, *args, **kwargs)
     64   def _method_wrapper(self, *args, **kwargs):
     65     if not self._in_multi_worker_mode():  # pylint: disable=protected-access
---> 66       return method(self, *args, **kwargs)
     67 
     68     # Running inside `run_distribute_coordinator` already.

/usr/local/lib/python3.6/dist-packages/tensorflow/python/keras/engine/training.py in fit(self, x, y, batch_size, epochs, verbose, callbacks, validation_split, validation_data, shuffle, class_weight, sample_weight, initial_epoch, steps_per_epoch, validation_steps, validation_batch_size, validation_freq, max_queue_size, workers, use_multiprocessing)
    846                 batch_size=batch_size):
    847               callbacks.on_train_batch_begin(step)
--> 848               tmp_logs = train_function(iterator)
    849               # Catch OutOfRangeError for Datasets of unknown size.
    850               # This blocks until the batch has finished executing.

/usr/local/lib/python3.6/dist-packages/tensorflow/python/eager/def_function.py in __call__(self, *args, **kwds)
    578         xla_context.Exit()
    579     else:
--> 580       result = self._call(*args, **kwds)
    581 
    582     if tracing_count == self._get_tracing_count():

/usr/local/lib/python3.6/dist-packages/tensorflow/python/eager/def_function.py in _call(self, *args, **kwds)
    642         # Lifting succeeded, so variables are initialized and we can run the
    643         # stateless function.
--> 644         return self._stateless_fn(*args, **kwds)
    645     else:
    646       canon_args, canon_kwds = \

/usr/local/lib/python3.6/dist-packages/tensorflow/python/eager/function.py in __call__(self, *args, **kwargs)
   2418     with self._lock:
   2419       graph_function, args, kwargs = self._maybe_define_function(args, kwargs)
-> 2420     return graph_function._filtered_call(args, kwargs)  # pylint: disable=protected-access
   2421 
   2422   @property

/usr/local/lib/python3.6/dist-packages/tensorflow/python/eager/function.py in _filtered_call(self, args, kwargs)
   1663          if isinstance(t, (ops.Tensor,
   1664                            resource_variable_ops.BaseResourceVariable))),
-> 1665         self.captured_inputs)
   1666 
   1667   def _call_flat(self, args, captured_inputs, cancellation_manager=None):

/usr/local/lib/python3.6/dist-packages/tensorflow/python/eager/function.py in _call_flat(self, args, captured_inputs, cancellation_manager)
   1744       # No tape is watching; skip to running the function.
   1745       return self._build_call_outputs(self._inference_function.call(
-> 1746           ctx, args, cancellation_manager=cancellation_manager))
   1747     forward_backward = self._select_forward_and_backward_functions(
   1748         args,

/usr/local/lib/python3.6/dist-packages/tensorflow/python/eager/function.py in call(self, ctx, args, cancellation_manager)
    596               inputs=args,
    597               attrs=attrs,
--> 598               ctx=ctx)
    599         else:
    600           outputs = execute.execute_with_cancellation(

/usr/local/lib/python3.6/dist-packages/tensorflow/python/eager/execute.py in quick_execute(op_name, num_outputs, inputs, attrs, ctx, name)
     58     ctx.ensure_initialized()
     59     tensors = pywrap_tfe.TFE_Py_Execute(ctx._handle, device_name, op_name,
---> 60                                         inputs, attrs, num_outputs)
     61   except core._NotOkStatusException as e:
     62     if name is not None:

InvalidArgumentError: Cannot assign a device for operation model/d_embedded/m1_embedding/m1_embedding_weights/embedding_lookup_sparse/embedding_lookup: Could not satisfy explicit device specification '' because the node {{colocation_node model/d_embedded/m1_embedding/m1_embedding_weights/embedding_lookup_sparse/embedding_lookup}} was colocated with a group of nodes that required incompatible device '/job:localhost/replica:0/task:0/device:GPU:0'. All available devices [/job:localhost/replica:0/task:0/device:CPU:0, /job:localhost/replica:0/task:0/device:XLA_CPU:0, /job:localhost/replica:0/task:0/device:XLA_GPU:0, /job:localhost/replica:0/task:0/device:GPU:0]. 
Colocation Debug Info:
Colocation group had the following types and supported devices: 
Root Member(assigned_device_name_index_=2 requested_device_name_='/job:localhost/replica:0/task:0/device:GPU:0' assigned_device_name_='/job:localhost/replica:0/task:0/device:GPU:0' resource_device_name_='/job:localhost/replica:0/task:0/device:GPU:0' supported_device_types_=[CPU] possible_devices_=[]
ResourceSparseApplyAdagradV2: CPU 
UnsortedSegmentSum: GPU CPU XLA_CPU XLA_GPU 
StridedSlice: GPU CPU XLA_CPU XLA_GPU 
Const: GPU CPU XLA_CPU XLA_GPU 
ResourceGather: GPU CPU XLA_CPU XLA_GPU 
Unique: GPU CPU 
Shape: GPU CPU XLA_CPU XLA_GPU 
_Arg: GPU CPU XLA_CPU XLA_GPU 
Identity: GPU CPU XLA_CPU XLA_GPU 
VariableShape: GPU CPU XLA_CPU XLA_GPU 

Colocation members, user-requested devices, and framework assigned devices, if any:
  model_d_embedded_m1_embedding_m1_embedding_weights_embedding_lookup_sparse_embedding_lookup_651 (_Arg)  framework assigned device=/job:localhost/replica:0/task:0/device:GPU:0
  adagrad_adagrad_update_resourcesparseapplyadagradv2_accum (_Arg)  framework assigned device=/job:localhost/replica:0/task:0/device:GPU:0
  model/d_embedded/m1_embedding/m1_embedding_weights/embedding_lookup_sparse/embedding_lookup (ResourceGather) 
  model/d_embedded/m1_embedding/m1_embedding_weights/embedding_lookup_sparse/embedding_lookup/Identity (Identity) 
  gradient_tape/model/d_embedded/m1_embedding/m1_embedding_weights/embedding_lookup_sparse/embedding_lookup/VariableShape (VariableShape) 
  Adagrad/Adagrad/update/Unique (Unique) 
  Adagrad/Adagrad/update/Shape (Shape) 
  Adagrad/Adagrad/update/strided_slice/stack (Const) 
  Adagrad/Adagrad/update/strided_slice/stack_1 (Const) 
  Adagrad/Adagrad/update/strided_slice/stack_2 (Const) 
  Adagrad/Adagrad/update/strided_slice (StridedSlice) 
  Adagrad/Adagrad/update/UnsortedSegmentSum (UnsortedSegmentSum) 
  Adagrad/Adagrad/update/ResourceSparseApplyAdagradV2 (ResourceSparseApplyAdagradV2) 

	 [[{{node model/d_embedded/m1_embedding/m1_embedding_weights/embedding_lookup_sparse/embedding_lookup}}]] [Op:__inference_train_function_913]
```

**Comment by kkranen** [comment]
Inside tensorflow/tensorflow:nightly-gpu container:

```
Colocation Debug Info:
Colocation group had the following types and supported devices:
Root Member(assigned_device_name_index_=2 requested_device_name_='/job:localhost/replica:0/task:0/device:GPU:0' assigned_device_name_='/job:localhost/replica:0/task:0/device:GPU:0' resource_device_name_='/job:localhost/replica:0/task:0/device:GPU:0' supported_device_types_=[CPU] possible_devices_=[]
ResourceSparseApplyAdagradV2: CPU
UnsortedSegmentSum: GPU CPU XLA_CPU XLA_GPU
StridedSlice: GPU CPU XLA_CPU XLA_GPU
Const: GPU CPU XLA_CPU XLA_GPU
ResourceGather: GPU CPU XLA_CPU XLA_GPU
Unique: GPU CPU
Shape: GPU CPU XLA_CPU XLA_GPU
_Arg: GPU CPU XLA_CPU XLA_GPU
Identity: GPU CPU XLA_CPU XLA_GPU
VariableShape: GPU CPU XLA_CPU XLA_GPU
Colocation members, user-requested devices, and framework assigned devices, if any:
  functional_1_d_embedded_m1_embedding_m1_embedding_weights_embedding_lookup_sparse_embedding_lookup_689 (_Arg)  framework assigned device=/job:localhost/replica:0/task:0/device:GPU:0
  adagrad_adagrad_update_resourcesparseapplyadagradv2_accum (_Arg)  framework assigned device=/job:localhost/replica:0/task:0/device:GPU:0
  functional_1/d_embedded/m1_embedding/m1_embedding_weights/embedding_lookup_sparse/embedding_lookup (ResourceGather) /job:localhost/replica:0/task:0/device:GPU:0
  functional_1/d_embedded/m1_embedding/m1_embedding_weights/embedding_lookup_sparse/embedding_lookup/Identity (Identity) /job:localhost/replica:0/task:0/device:GPU:0
  gradient_tape/functional_1/d_embedded/m1_embedding/m1_embedding_weights/embedding_lookup_sparse/embedding_lookup/VariableShape (VariableShape)
  Adagrad/Adagrad/update/Unique (Unique) /job:localhost/replica:0/task:0/device:GPU:0
  Adagrad/Adagrad/update/Shape (Shape) /job:localhost/replica:0/task:0/device:GPU:0
  Adagrad/Adagrad/update/strided_slice/stack (Const) /job:localhost/replica:0/task:0/device:GPU:0
  Adagrad/Adagrad/update/strided_slice/stack_1 (Const) /job:localhost/replica:0/task:0/device:GPU:0
  Adagrad/Adagrad/update/strided_slice/stack_2 (Const) /job:localhost/replica:0/task:0/device:GPU:0
  Adagrad/Adagrad/update/strided_slice (StridedSlice) /job:localhost/replica:0/task:0/device:GPU:0
  Adagrad/Adagrad/update/UnsortedSegmentSum (UnsortedSegmentSum) /job:localhost/replica:0/task:0/device:GPU:0
  Adagrad/Adagrad/update/ResourceSparseApplyAdagradV2 (ResourceSparseApplyAdagradV2) /job:localhost/replica:0/task:0/device:GPU:0
Op: ResourceGather
Node attrs: dtype=DT_FLOAT, batch_dims=0, Tindices=DT_INT64, validate_indices=true, _class=["loc:@funct...lookup/689"]
Registered kernels:
  device='XLA_GPU'; Tindices in [DT_INT32, DT_INT64]; dtype in [DT_FLOAT, DT_DOUBLE, DT_INT32, DT_UINT8, DT_INT16, DT_INT8, DT_COMPLEX64, DT_INT64, DT_BOOL, DT_QINT8, DT_QUINT8, DT_QINT32, DT_BFLOAT16, DT_UINT16, DT_COMPLEX128, DT_HALF, DT_UINT32, DT_UINT64]
  device='CPU'; dtype in [DT_QINT32]; Tindices in [DT_INT64]
  device='CPU'; dtype in [DT_QINT32]; Tindices in [DT_INT32]
  device='CPU'; dtype in [DT_QUINT8]; Tindices in [DT_INT64]
  device='CPU'; dtype in [DT_QUINT8]; Tindices in [DT_INT32]
  device='CPU'; dtype in [DT_QINT8]; Tindices in [DT_INT64]
  device='CPU'; dtype in [DT_QINT8]; Tindices in [DT_INT32]
  device='CPU'; dtype in [DT_VARIANT]; Tindices in [DT_INT64]
  device='CPU'; dtype in [DT_VARIANT]; Tindices in [DT_INT32]
  device='CPU'; dtype in [DT_RESOURCE]; Tindices in [DT_INT64]
  device='CPU'; dtype in [DT_RESOURCE]; Tindices in [DT_INT32]
  device='CPU'; dtype in [DT_STRING]; Tindices in [DT_INT64]
  device='CPU'; dtype in [DT_STRING]; Tindices in [DT_INT32]
  device='CPU'; dtype in [DT_BOOL]; Tindices in [DT_INT64]
  device='CPU'; dtype in [DT_BOOL]; Tindices in [DT_INT32]
  device='CPU'; dtype in [DT_COMPLEX128]; Tindices in [DT_INT64]
  device='CPU'; dtype in [DT_COMPLEX128]; Tindices in [DT_INT32]
  device='CPU'; dtype in [DT_COMPLEX64]; Tindices in [DT_INT64]
  device='CPU'; dtype in [DT_COMPLEX64]; Tindices in [DT_INT32]
  device='CPU'; dtype in [DT_DOUBLE]; Tindices in [DT_INT64]
  device='CPU'; dtype in [DT_DOUBLE]; Tindices in [DT_INT32]
  device='CPU'; dtype in [DT_FLOAT]; Tindices in [DT_INT64]
  device='CPU'; dtype in [DT_FLOAT]; Tindices in [DT_INT32]
  device='CPU'; dtype in [DT_BFLOAT16]; Tindices in [DT_INT64]
  device='CPU'; dtype in [DT_BFLOAT16]; Tindices in [DT_INT32]
  device='CPU'; dtype in [DT_HALF]; Tindices in [DT_INT64]
  device='CPU'; dtype in [DT_HALF]; Tindices in [DT_INT32]
  device='CPU'; dtype in [DT_INT8]; Tindices in [DT_INT64]
  device='CPU'; dtype in [DT_INT8]; Tindices in [DT_INT32]
  device='CPU'; dtype in [DT_UINT8]; Tindices in [DT_INT64]
  device='CPU'; dtype in [DT_UINT8]; Tindices in [DT_INT32]
  device='CPU'; dtype in [DT_INT16]; Tindices in [DT_INT64]
  device='CPU'; dtype in [DT_INT16]; Tindices in [DT_INT32]
  device='CPU'; dtype in [DT_UINT16]; Tindices in [DT_INT64]
  device='CPU'; dtype in [DT_UINT16]; Tindices in [DT_INT32]
  device='CPU'; dtype in [DT_INT32]; Tindices in [DT_INT64]
  device='CPU'; dtype in [DT_INT32]; Tindices in [DT_INT32]
  device='CPU'; dtype in [DT_UINT32]; Tindices in [DT_INT64]
  device='CPU'; dtype in [DT_UINT32]; Tindices in [DT_INT32]
  device='CPU'; dtype in [DT_INT64]; Tindices in [DT_INT64]
  device='CPU'; dtype in [DT_INT64]; Tindices in [DT_INT32]
  device='CPU'; dtype in [DT_UINT64]; Tindices in [DT_INT64]
  device='CPU'; dtype in [DT_UINT64]; Tindices in [DT_INT32]
  device='GPU'; dtype in [DT_VARIANT]; Tindices in [DT_INT64]
  device='GPU'; dtype in [DT_VARIANT]; Tindices in [DT_INT32]
  device='GPU'; dtype in [DT_BOOL]; Tindices in [DT_INT64]
  device='GPU'; dtype in [DT_BOOL]; Tindices in [DT_INT32]
  device='GPU'; dtype in [DT_COMPLEX128]; Tindices in [DT_INT64]
  device='GPU'; dtype in [DT_COMPLEX128]; Tindices in [DT_INT32]
  device='GPU'; dtype in [DT_COMPLEX64]; Tindices in [DT_INT64]
  device='GPU'; dtype in [DT_COMPLEX64]; Tindices in [DT_INT32]
  device='GPU'; dtype in [DT_DOUBLE]; Tindices in [DT_INT64]
  device='GPU'; dtype in [DT_DOUBLE]; Tindices in [DT_INT32]
  device='GPU'; dtype in [DT_FLOAT]; Tindices in [DT_INT64]
  device='GPU'; dtype in [DT_FLOAT]; Tindices in [DT_INT32]
  device='GPU'; dtype in [DT_HALF]; Tindices in [DT_INT64]
  device='GPU'; dtype in [DT_HALF]; Tindices in [DT_INT32]
  device='GPU'; dtype in [DT_INT64]; Tindices in [DT_INT64]
  device='GPU'; dtype in [DT_INT64]; Tindices in [DT_INT32]
  device='XLA_GPU_JIT'; Tindices in [DT_INT32, DT_INT64]; dtype in [DT_FLOAT, DT_DOUBLE, DT_INT32, DT_UINT8, DT_INT16, DT_INT8, DT_COMPLEX64, DT_INT64, DT_BOOL, DT_QINT8, DT_QUINT8, DT_QINT32, DT_BFLOAT16, DT_UINT16, DT_COMPLEX128, DT_HALF, DT_UINT32, DT_UINT64]
  device='XLA_CPU_JIT'; Tindices in [DT_INT32, DT_INT64]; dtype in [DT_FLOAT, DT_DOUBLE, DT_INT32, DT_UINT8, DT_INT16, DT_INT8, DT_COMPLEX64, DT_INT64, DT_BOOL, DT_QINT8, DT_QUINT8, DT_QINT32, DT_BFLOAT16, DT_UINT16, DT_COMPLEX128, DT_HALF, DT_UINT32, DT_UINT64]
  device='XLA_CPU'; Tindices in [DT_INT32, DT_INT64]; dtype in [DT_FLOAT, DT_DOUBLE, DT_INT32, DT_UINT8, DT_INT16, DT_INT8, DT_COMPLEX64, DT_INT64, DT_BOOL, DT_QINT8, DT_QUINT8, DT_QINT32, DT_BFLOAT16, DT_UINT16, DT_COMPLEX128, DT_HALF, DT_UINT32, DT_UINT64]
         [[{{node functional_1/d_embedded/m1_embedding/m1_embedding_weights/embedding_lookup_sparse/embedding_lookup}}]] [Op:__inference_train_function_952]
```

**Comment by jvishnuvardhan** [comment]
@kkranen I can reproduce the issue when I select `GPU` in the colab. [Here](https://colab.research.google.com/gist/jvishnuvardhan/5d41769e7eb3aa69e70fb70c53d97fb7/untitled300.ipynb) is the gist for our reference. Thanks!

**Comment by sanjoy** [comment]
Is it correct to say that the root cause is that `SparseApplyAdagradV2` does not have a GPU kernel?

**Comment by DEKHTIARJonathan** [comment]
It seems the case, that would also be my bet: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/training_ops.cc#L1938-L1948

**Comment by tanzhenyu** [comment]
Quickly reading the issue, why is this related to `DenseFeatures` -- it will also error out w/o it, correct?

**Comment by kkranen** [comment]
Initially I thought this was related to dense features but it seems to error without it.

**Comment by tanzhenyu** [comment]
> Initially I thought this was related to dense features but it seems to error without it.

That'd my guess as well. And yes missing GPU kernels for those sparse ops is the main reason I believe.

**Comment by sushreebarsa** [comment]
Was able to reproduce the issue in TF v2.5 ,please find the gist [here](https://colab.research.google.com/gist/sushreebarsa/5801f85168ed86a1eaa6bfcb5b2a95cf/untitled139.ipynb?authuser=1)..Thanks !

**Comment by benbarsdell** [comment]
I haven't tried reproducing this issue, but I'll note that a GPU kernel for SparseApplyAdagradV2 was [added](https://github.com/tensorflow/tensorflow/commit/28162ecc0cf0993fa72e4d3b72f864300188ccd5) in TF 2.5.

**Comment by benbarsdell** [comment]
I ran the gist on Colab and locally on my machine with TF 2.5 and 2.6 and in all cases it worked without error. I do see the error with TF 2.4. So afaict this issue is fixed as of TF 2.5.
@sushreebarsa can you confirm that you reproduced the error with TF 2.5?

**Comment by Saduf2019** [comment]
@benbarsdell 
I ran the code on tf 2.5 and it works as expected, please find the [gist here](https://colab.research.google.com/gist/Saduf2019/d8d47fc4b28fc5cc64b48b74388800a0/untitled139.ipynb?authuser=1), can you please move this to closed status as its resolved.

**Comment by benbarsdell** [comment]
Closing as fixed since TF 2.5.
